homeassistant/components/here_weather/__init__.py

```diff
@@ -0,0 +1,187 @@
+        self.coordinator = None
+        self.unsub_handler = None
+
+    async def async_setup(self):
+        """Set up the here_weather integration."""
+        self.add_options()
+        self.unsub_handler = self.config_entry.add_update_listener(
+            self.async_options_updated
+        )
+        await self._async_create_coordinator()
+        return True
+
+    def add_options(self):
+        """Add options for here_weather integration."""
+        if not self.config_entry.options:
+            options = {CONF_SCAN_INTERVAL: DEFAULT_SCAN_INTERVAL}
+            self.hass.config_entries.async_update_entry(
+                self.config_entry, options=options
+            )
+
+    async def _async_create_coordinator(self):
+        """Create or recreate the DataUpdateCoordinator."""
+        self.coordinator = DataUpdateCoordinator(
+            self.hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_method=self.async_update,
+            update_interval=timedelta(
+                seconds=self.config_entry.options[CONF_SCAN_INTERVAL]
+            ),
+        )
+        await self.coordinator.async_refresh()
+
+    async def async_update(self) -> None:
+        """Handle data update with the DataUpdateCoordinator."""
+        try:
+            async with async_timeout.timeout(10):
+                return await self.hass.async_add_executor_job(self._get_data)
+        except herepy.InvalidRequestError as error:
+            raise UpdateFailed(f"Unable to fetch data from HERE: {error.message}")
+
+    def _get_data(self):
+        """Get the latest data from HERE."""
+        is_metric = convert_units_to_boolean(self.units)
+        if self.zip_code is not None:
+            data = self.here_client.weather_for_zip_code(
+                self.zip_code, self.weather_product_type, metric=is_metric
+            )
+        elif self.location_name is not None:
+            data = self.here_client.weather_for_location_name(
+                self.location_name, self.weather_product_type, metric=is_metric
+            )
+        else:
+            data = self.here_client.weather_for_coordinates(
+                self.latitude,
+                self.longitude,
+                self.weather_product_type,
+                metric=is_metric,
+            )
+        return extract_data_from_payload_for_product_type(
+            data, self.weather_product_type
+        )
+
+    @staticmethod
+    async def async_options_updated(
+        hass: HomeAssistant, config_entry: ConfigEntry
+    ) -> None:
+        """Triggered by config entry options updates."""
+        await hass.data[DOMAIN][config_entry.entry_id].async_set_scan_interval(
+            config_entry.options[CONF_SCAN_INTERVAL]
+        )
+
+    async def async_set_scan_interval(self, scan_interval):
+        """Recreate the coordinator with the new scan interval."""
+        await self._async_create_coordinator()
```

Comment:
When are the previous coordinator's listeners removed? This seems like the old coordinator would remain running and updating listeners as it syncs.